File: src_ts/reviewer.ts
import { normalizeConfig, type AdapterResult, type ScanConfigInput } from "./config.js";
import { scanText } from "./index.js";
import type { Contradiction, DetectionReport, Finding } from "./types.js";

export type ReviewDecisionValue = "approve" | "block" | "review";

export interface ReviewDecision {
  decision: ReviewDecisionValue;
  reason: string;
  report: DetectionReport;
  reviewerResult?: AdapterResult;
}

export function reviewTextWrite(text: string, config: ScanConfigInput = {}): ReviewDecision {
  const normalized = normalizeConfig(config);
  // Reviewer needs evidence; the returned report is stripped before exiting.
  const fullReport = scanText(text, { includeEvidence: true, config: normalized });
  const strippedReport = stripEvidence(fullReport);
  if (!fullReport.isPromptInjection && !fullReport.hasPii && !fullReport.isContradictory) {
    return { decision: "approve", reason: "No configured detector crossed its threshold.", report: strippedReport };
  }
  if (!normalized.reviewerAdapter) {
    return {
      decision: "review",
      reason: "Local scan crossed a threshold; no reviewer adapter is configured.",
      report: strippedReport,
    };
  }
  const reviewerResult = normalized.reviewerAdapter(text, fullReport);
  const blockThreshold = normalized.thresholds.reviewerBlock;
  if (reviewerResult.score >= blockThreshold) {
    return { decision: "block", reason: reviewerResult.label, report: strippedReport, reviewerResult };
  }
  return { decision: "approve", reason: reviewerResult.label, report: strippedReport, reviewerResult };
}

export async function reviewTextWriteAsync(
  text: string,
  config: ScanConfigInput = {},
): Promise<ReviewDecision> {
  const normalized = normalizeConfig(config);
  const fullReport = scanText(text, { includeEvidence: true, config: normalized });
  const strippedReport = stripEvidence(fullReport);
  if (!fullReport.isPromptInjection && !fullReport.hasPii && !fullReport.isContradictory) {
    return { decision: "approve", reason: "No configured detector crossed its threshold.", report: strippedReport };
  }
  let reviewerResult: AdapterResult | undefined;
  if (normalized.asyncReviewerAdapter) {
    reviewerResult = await normalized.asyncReviewerAdapter(text, fullReport);
  } else if (normalized.reviewerAdapter) {
    reviewerResult = normalized.reviewerAdapter(text, fullReport);
  }
  if (!reviewerResult) {
    return {
      decision: "review",
      reason: "Local scan crossed a threshold; no reviewer adapter is configured.",
      report: strippedReport,
    };
  }
  const blockThreshold = normalized.thresholds.reviewerBlock;
  if (reviewerResult.score >= blockThreshold) {
    return { decision: "block", reason: reviewerResult.label, report: strippedReport, reviewerResult };
  }
  return { decision: "approve", reason: reviewerResult.label, report: strippedReport, reviewerResult };
}

function stripEvidence(report: DetectionReport): DetectionReport {
  return {
    ...report,
    promptInjectionFindings: report.promptInjectionFindings.map(stripFinding),
    piiFindings: report.piiFindings.map(stripFinding),
    contradictions: report.contradictions.map(stripContradiction),
  };
}

function stripFinding(finding: Finding): Finding {
  return { ...finding, evidence: "" };
}

function stripContradiction(item: Contradiction): Contradiction {
  return { ...item, positive: "", negative: "" };
}

File: src_py/prompt_guard/reviewer.py
from __future__ import annotations

from dataclasses import dataclass, replace
from typing import Literal, Optional

from .config import DEFAULT_CONFIG, AdapterResult, ScanConfig
from .core import scan_text
from .types import DetectionReport


Decision = Literal["approve", "block", "review"]


@dataclass(frozen=True)
class ReviewDecision:
    decision: Decision
    reason: str
    report: DetectionReport
    reviewer_result: Optional[AdapterResult] = None


def review_text_write(text: str, config: ScanConfig = DEFAULT_CONFIG) -> ReviewDecision:
    # Reviewer needs evidence to make a decision; the returned report is stripped.
    full_report = scan_text(text, include_evidence=True, config=config)
    stripped_report = _strip_evidence(full_report)
    if not (full_report.is_prompt_injection or full_report.has_pii or full_report.is_contradictory):
        return ReviewDecision("approve", "No configured detector crossed its threshold.", stripped_report)
    if config.reviewer_adapter is None:
        return ReviewDecision(
            "review",
            "Local scan crossed a threshold; no reviewer adapter is configured.",
            stripped_report,
        )
    reviewer_result = config.reviewer_adapter(text, full_report)
    block_threshold = config.thresholds.reviewer_block
    if reviewer_result.score >= block_threshold:
        return ReviewDecision("block", reviewer_result.label, stripped_report, reviewer_result)
    return ReviewDecision("approve", reviewer_result.label, stripped_report, reviewer_result)


async def review_text_write_async(text: str, config: ScanConfig = DEFAULT_CONFIG) -> ReviewDecision:
    full_report = scan_text(text, include_evidence=True, config=config)
    stripped_report = _strip_evidence(full_report)
    if not (full_report.is_prompt_injection or full_report.has_pii or full_report.is_contradictory):
        return ReviewDecision("approve", "No configured detector crossed its threshold.", stripped_report)
    reviewer_result: Optional[AdapterResult] = None
    if config.async_reviewer_adapter is not None:
        reviewer_result = await config.async_reviewer_adapter(text, full_report)
    elif config.reviewer_adapter is not None:
        reviewer_result = config.reviewer_adapter(text, full_report)
    if reviewer_result is None:
        return ReviewDecision(
            "review",
            "Local scan crossed a threshold; no reviewer adapter is configured.",
            stripped_report,
        )
    block_threshold = config.thresholds.reviewer_block
    if reviewer_result.score >= block_threshold:
        return ReviewDecision("block", reviewer_result.label, stripped_report, reviewer_result)
    return ReviewDecision("approve", reviewer_result.label, stripped_report, reviewer_result)


def _strip_evidence(report: DetectionReport) -> DetectionReport:
    return replace(
        report,
        prompt_injection_findings=[replace(f, evidence="") for f in report.prompt_injection_findings],
        pii_findings=[replace(f, evidence="") for f in report.pii_findings],
        contradictions=[replace(item, positive="", negative="") for item in report.contradictions],
    )
